File: java/src/leetcode/p062/Solution.java
package leetcode.p062;

import java.util.Iterator;
import java.util.*;


class Solution {
    public static void main(String[] args) {
        Solution s = new Solution();
        int res = s.uniquePaths(30, 30);
        System.out.println(res);
    }

    private int solution = 0;
    private int[][] cache = null;


    public int uniquePaths(int m, int n) {
        solution = 0;
        cache = new int[m][n];
        return dynmove(m - 1, n - 1);
    }

    private int dynmove(int m, int n) {
        if (m == 0 && n == 0)
            return 1;
        else {
            if (cache[m][n] != 0)
                return cache[m][n];
            else {
                int top = 0;
                int left = 0;
                if (m > 0)
                    if (cache[m - 1][n] != 0)
                        top = cache[m - 1][n];
                    else
                        top = dynmove(m - 1, n);
                if (n > 0)
                    if (cache[m][n - 1] != 0)
                        left = cache[m][n - 1];
                    else
                        left = dynmove(m, n - 1);
                cache[m][n] = top + left;
                return top + left;
            }
        }
    }

    private void move(int m, int n, int posx, int posy) {
        if (m == posx && n == posy) {
            solution++;
        } else if (m == posx)
            move(m, n, posx, posy + 1);
        else if (n == posy)
            move(m, n, posx + 1, posy);
        else {
            move(m, n, posx, posy + 1);
            move(m, n, posx + 1, posy);
        }
    }


}

File: scala/src/leetcode/p062/Solution.scala
package leetcode.p062


object Solution {
    def uniquePaths(m: Int, n: Int): Int = {
       val arr = Array.ofDim[Int](m,n)
       for (j <- 0 until n){
        arr(0)(j) = 1
       }
       for (i <- 0 until m){
        arr(i)(0) = 1
       }
       for (i <- 1 until m){
        for (j <- 1 until n){
            val up = arr(i-1)(j)
            val left = arr(i)(j-1)
            arr(i)(j) = up + left 
        }
       }
       return arr(m-1)(n-1)
    }
}